resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/KubernetesExternalShuffleService.scala

```
@@ -0,0 +1,133 @@
+ */
+
+package org.apache.spark.deploy.kubernetes
+
+import java.nio.ByteBuffer
+
+import scala.collection.JavaConverters._
+
+import io.fabric8.kubernetes.api.model.Pod
+import io.fabric8.kubernetes.client.{KubernetesClientException, Watch, Watcher}
+import io.fabric8.kubernetes.client.Watcher.Action
+
+import org.apache.spark.{SecurityManager, SparkConf}
+import org.apache.spark.deploy.ExternalShuffleService
+import org.apache.spark.deploy.kubernetes.constants._
+import org.apache.spark.internal.Logging
+import org.apache.spark.network.client.{RpcResponseCallback, TransportClient}
+import org.apache.spark.network.shuffle.ExternalShuffleBlockHandler
+import org.apache.spark.network.shuffle.protocol.BlockTransferMessage
+import org.apache.spark.network.shuffle.protocol.kubernetes.RegisterKubernetesApp
+import org.apache.spark.network.util.TransportConf
+import org.apache.spark.scheduler.cluster.kubernetes.DriverPodKubernetesClientProvider
+
+
+/**
+ * An RPC endpoint that receives registration requests from Spark drivers running on Kubernetes.
+ * It detects driver termination and calls the cleanup callback to [[ExternalShuffleService]].
+ */
+private[spark] class KubernetesShuffleBlockHandler (
+  sparkConf: SparkConf,
+  transportConf: TransportConf)
+  extends ExternalShuffleBlockHandler(transportConf, null) with Logging {
+
+  private var connectedApps: Set[String] = Set()
+  private val CONNECTED_APPS_LOCK = new Object
+  private val Watch: Option[Watch] = ShuffleWatcher()
+
+  protected override def handleMessage(
+    message: BlockTransferMessage,
+    client: TransportClient,
+    callback: RpcResponseCallback): Unit = {
+      message match {
+        case RegisterDriverParam(appId) =>
+          val address = client.getSocketAddress
+          logInfo(s"Received registration request from app $appId (remote address $address).")
+          CONNECTED_APPS_LOCK.synchronized {
+            if (connectedApps.contains(appId)) {
```

Do we want to throw an error here instead of just logging the warning? I'm not sure what the implications are if we throw an error here.

---

It's actually not an error case, because when multiple executors are launched on the same node, they may register with the same shuffle service.